File: sdk/python/src/litegraph_sdk/models/authorization.py
import uuid
from datetime import datetime, timezone
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class AuthorizationRoleModel(BaseModel):
    """
    Authorization role metadata.
    """

    guid: str = Field(default_factory=lambda: str(uuid.uuid4()), alias="GUID")
    tenant_guid: Optional[str] = Field(default=None, alias="TenantGUID")
    name: str = Field(default="", alias="Name")
    display_name: Optional[str] = Field(default=None, alias="DisplayName")
    description: Optional[str] = Field(default=None, alias="Description")
    built_in: bool = Field(default=False, alias="BuiltIn")
    built_in_role: Optional[str] = Field(default="Custom", alias="BuiltInRole")
    resource_scope: str = Field(default="Tenant", alias="ResourceScope")
    permissions: List[str] = Field(default_factory=list, alias="Permissions")
    resource_types: List[str] = Field(default_factory=list, alias="ResourceTypes")
    inherits_to_graphs: bool = Field(default=False, alias="InheritsToGraphs")
    created_utc: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc), alias="CreatedUtc"
    )
    last_update_utc: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc), alias="LastUpdateUtc"
    )

    model_config = ConfigDict(populate_by_name=True, from_attributes=True)


class UserRoleAssignmentModel(BaseModel):
    """
    User role assignment.
    """

    guid: str = Field(default_factory=lambda: str(uuid.uuid4()), alias="GUID")
    tenant_guid: str = Field(alias="TenantGUID")
    user_guid: str = Field(alias="UserGUID")
    role_guid: Optional[str] = Field(default=None, alias="RoleGUID")
    role_name: Optional[str] = Field(default=None, alias="RoleName")
    resource_scope: str = Field(default="Tenant", alias="ResourceScope")
    graph_guid: Optional[str] = Field(default=None, alias="GraphGUID")
    created_utc: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc), alias="CreatedUtc"
    )
    last_update_utc: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc), alias="LastUpdateUtc"
    )

    model_config = ConfigDict(populate_by_name=True, from_attributes=True)


class CredentialScopeAssignmentModel(BaseModel):
    """
    Credential scope assignment.
    """

    guid: str = Field(default_factory=lambda: str(uuid.uuid4()), alias="GUID")
    tenant_guid: str = Field(alias="TenantGUID")
    credential_guid: str = Field(alias="CredentialGUID")
    role_guid: Optional[str] = Field(default=None, alias="RoleGUID")
    role_name: Optional[str] = Field(default=None, alias="RoleName")
    resource_scope: str = Field(default="Tenant", alias="ResourceScope")
    graph_guid: Optional[str] = Field(default=None, alias="GraphGUID")
    permissions: List[str] = Field(default_factory=list, alias="Permissions")
    resource_types: List[str] = Field(default_factory=list, alias="ResourceTypes")
    created_utc: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc), alias="CreatedUtc"
    )
    last_update_utc: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc), alias="LastUpdateUtc"
    )

    model_config = ConfigDict(populate_by_name=True, from_attributes=True)


class AuthorizationEffectiveGrantModel(BaseModel):
    """
    Effective authorization grant resolved for a user or credential.
    """

    source: str = Field(alias="Source")
    assignment_guid: str = Field(alias="AssignmentGUID")
    role_guid: Optional[str] = Field(default=None, alias="RoleGUID")
    role_name: Optional[str] = Field(default=None, alias="RoleName")
    resource_scope: str = Field(alias="ResourceScope")
    graph_guid: Optional[str] = Field(default=None, alias="GraphGUID")
    permissions: List[str] = Field(default_factory=list, alias="Permissions")
    resource_types: List[str] = Field(default_factory=list, alias="ResourceTypes")
    inherits_to_graphs: bool = Field(default=False, alias="InheritsToGraphs")
    applies_to_requested_graph: bool = Field(
        default=False, alias="AppliesToRequestedGraph"
    )

    model_config = ConfigDict(populate_by_name=True, from_attributes=True)


class AuthorizationEffectivePermissionsResultModel(BaseModel):
    """
    Effective authorization permissions for a user or credential.
    """

    tenant_guid: str = Field(alias="TenantGUID")
    user_guid: Optional[str] = Field(default=None, alias="UserGUID")
    credential_guid: Optional[str] = Field(default=None, alias="CredentialGUID")
    graph_guid: Optional[str] = Field(default=None, alias="GraphGUID")
    grants: List[AuthorizationEffectiveGrantModel] = Field(
        default_factory=list, alias="Grants"
    )
    user_role_assignments: List[UserRoleAssignmentModel] = Field(
        default_factory=list, alias="UserRoleAssignments"
    )
    credential_scope_assignments: List[CredentialScopeAssignmentModel] = Field(
        default_factory=list, alias="CredentialScopeAssignments"
    )
    roles: List[AuthorizationRoleModel] = Field(default_factory=list, alias="Roles")

    model_config = ConfigDict(populate_by_name=True, from_attributes=True)


class AuthorizationRoleSearchResultModel(BaseModel):
    """
    Authorization role search result.
    """

    objects: List[AuthorizationRoleModel] = Field(default_factory=list, alias="Objects")
    page: int = Field(default=0, alias="Page")
    page_size: int = Field(default=100, alias="PageSize")
    total_count: int = Field(default=0, alias="TotalCount")
    total_pages: int = Field(default=0, alias="TotalPages")

    model_config = ConfigDict(populate_by_name=True, from_attributes=True)


class UserRoleAssignmentSearchResultModel(BaseModel):
    """
    User role assignment search result.
    """

    objects: List[UserRoleAssignmentModel] = Field(
        default_factory=list, alias="Objects"
    )
    page: int = Field(default=0, alias="Page")
    page_size: int = Field(default=100, alias="PageSize")
    total_count: int = Field(default=0, alias="TotalCount")
    total_pages: int = Field(default=0, alias="TotalPages")

    model_config = ConfigDict(populate_by_name=True, from_attributes=True)


class CredentialScopeAssignmentSearchResultModel(BaseModel):
    """
    Credential scope assignment search result.
    """

    objects: List[CredentialScopeAssignmentModel] = Field(
        default_factory=list, alias="Objects"
    )
    page: int = Field(default=0, alias="Page")
    page_size: int = Field(default=100, alias="PageSize")
    total_count: int = Field(default=0, alias="TotalCount")
    total_pages: int = Field(default=0, alias="TotalPages")

    model_config = ConfigDict(populate_by_name=True, from_attributes=True)

File: dashboard/src/lib/sdk/authorization.ts
import { sdk } from './litegraph.service';

export type AuthorizationPermission = 'Read' | 'Write' | 'Delete' | 'Admin';
export type AuthorizationResourceScope = 'Tenant' | 'Graph';
export type AuthorizationResourceType =
  | 'Admin'
  | 'Graph'
  | 'Node'
  | 'Edge'
  | 'Label'
  | 'Tag'
  | 'Vector'
  | 'Query'
  | 'Transaction';
export type BuiltInRole = 'TenantAdmin' | 'GraphAdmin' | 'Editor' | 'Viewer' | 'Custom';

export type AuthorizationRole = {
  GUID: string;
  TenantGUID?: string | null;
  Name: string;
  DisplayName?: string | null;
  Description?: string | null;
  BuiltIn: boolean;
  BuiltInRole?: BuiltInRole;
  ResourceScope: AuthorizationResourceScope;
  Permissions: AuthorizationPermission[];
  ResourceTypes: AuthorizationResourceType[];
  InheritsToGraphs: boolean;
  CreatedUtc?: string;
  LastUpdateUtc?: string;
};

export type UserRoleAssignment = {
  GUID: string;
  TenantGUID: string;
  UserGUID: string;
  RoleGUID?: string | null;
  RoleName?: string | null;
  ResourceScope: AuthorizationResourceScope;
  GraphGUID?: string | null;
  CreatedUtc?: string;
  LastUpdateUtc?: string;
};

export type CredentialScopeAssignment = {
  GUID: string;
  TenantGUID: string;
  CredentialGUID: string;
  RoleGUID?: string | null;
  RoleName?: string | null;
  ResourceScope: AuthorizationResourceScope;
  GraphGUID?: string | null;
  Permissions: AuthorizationPermission[];
  ResourceTypes: AuthorizationResourceType[];
  CreatedUtc?: string;
  LastUpdateUtc?: string;
};

export type AuthorizationEffectiveGrant = {
  Source: string;
  AssignmentGUID: string;
  RoleGUID?: string | null;
  RoleName?: string | null;
  ResourceScope: AuthorizationResourceScope;
  GraphGUID?: string | null;
  Permissions: AuthorizationPermission[];
  ResourceTypes: AuthorizationResourceType[];
  InheritsToGraphs: boolean;
  AppliesToRequestedGraph: boolean;
};

export type AuthorizationEffectivePermissionsResult = {
  TenantGUID: string;
  UserGUID?: string | null;
  CredentialGUID?: string | null;
  GraphGUID?: string | null;
  Grants: AuthorizationEffectiveGrant[];
  UserRoleAssignments: UserRoleAssignment[];
  CredentialScopeAssignments: CredentialScopeAssignment[];
  Roles: AuthorizationRole[];
};

export type AuthorizationSearchResult<T> = {
  Objects: T[];
  Page: number;
  PageSize: number;
  TotalCount: number;
  TotalPages: number;
};

export type AuthorizationListParams = {
  page?: number;
  pageSize?: number;
};

export type RoleListParams = AuthorizationListParams & {
  includeBuiltIns?: boolean;
  builtIn?: boolean;
};

const getBaseUrl = (): string => {
  const endpoint = sdk.config.endpoint || '/';
  return endpoint.endsWith('/') ? endpoint.slice(0, -1) : endpoint;
};

const buildHeaders = (): Record<string, string> => {
  const headers: Record<string, string> = {
    Accept: 'application/json',
  };
  const defaults = (sdk.config as unknown as { defaultHeaders?: Record<string, string> })
    .defaultHeaders;
  if (defaults) {
    for (const key of Object.keys(defaults)) headers[key] = defaults[key];
  }
  const authConfig = sdk.config as unknown as { accessToken?: string; accessKey?: string };
  const bearerToken = authConfig.accessToken || authConfig.accessKey;
  if (bearerToken && !headers.Authorization) {
    headers.Authorization = `Bearer ${bearerToken}`;
  }
  return headers;
};

const buildQuery = (params: Record<string, string | number | boolean | undefined>): string => {
  const entries = Object.entries(params).filter(
    ([, value]) => value !== undefined && value !== null && value !== ''
  );
  if (entries.length === 0) return '';
  return (
    '?' +
    entries
      .map(([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(String(value))}`)
      .join('&')
  );
};

const request = async <T>(method: string, url: string, body?: unknown): Promise<T> => {
  const headers = buildHeaders();
  if (body !== undefined) headers['Content-Type'] = 'application/json';
  const response = await fetch(url, {
    method,
    headers,
    body: body === undefined ? undefined : JSON.stringify(body),
  });
  if (!response.ok) {
    let message = `HTTP ${response.status} ${response.statusText}`;
    try {
      const errorBody = await response.json();
      message = errorBody?.Description || errorBody?.Message || message;
    } catch {
      // Keep the HTTP status message when the server did not return JSON.
    }
    throw new Error(message);
  }
  if (response.status === 204) return undefined as T;
  const text = await response.text();
  if (!text) return undefined as T;
  return JSON.parse(text) as T;
};

export const listAuthorizationRoles = (tenantGuid: string, params: RoleListParams = {}) => {
  const url = `${getBaseUrl()}/v1.0/tenants/${encodeURIComponent(tenantGuid)}/roles${buildQuery({
    page: params.page ?? 0,
    pageSize: params.pageSize ?? 1000,
    includeBuiltIns: params.includeBuiltIns ?? true,
    builtIn: params.builtIn,
  })}`;
  return request<AuthorizationSearchResult<AuthorizationRole>>('GET', url);
};

export const createAuthorizationRole = (tenantGuid: string, role: Partial<AuthorizationRole>) => {
  const url = `${getBaseUrl()}/v1.0/tenants/${encodeURIComponent(tenantGuid)}/roles`;
  return request<AuthorizationRole>('PUT', url, role);
};

export const updateAuthorizationRole = (tenantGuid: string, role: AuthorizationRole) => {
  const url = `${getBaseUrl()}/v1.0/tenants/${encodeURIComponent(
    tenantGuid
  )}/roles/${encodeURIComponent(role.GUID)}`;
  return request<AuthorizationRole>('PUT', url, role);
};

export const deleteAuthorizationRole = (tenantGuid: string, roleGuid: string) => {
  const url = `${getBaseUrl()}/v1.0/tenants/${encodeURIComponent(
    tenantGuid
  )}/roles/${encodeURIComponent(roleGuid)}`;
  return request<void>('DELETE', url);
};

export const listUserRoleAssignments = (
  tenantGuid: string,
  userGuid: string,
  params: AuthorizationListParams = {}
) => {
  const url = `${getBaseUrl()}/v1.0/tenants/${encodeURIComponent(
    tenantGuid
  )}/users/${encodeURIComponent(userGuid)}/roles${buildQuery({
    page: params.page ?? 0,
    pageSize: params.pageSize ?? 1000,
  })}`;
  return request<AuthorizationSearchResult<UserRoleAssignment>>('GET', url);
};

export const createUserRoleAssignment = (
  tenantGuid: string,
  userGuid: string,
  assignment: Partial<UserRoleAssignment>
) => {
  const url = `${getBaseUrl()}/v1.0/tenants/${encodeURIComponent(
    tenantGuid
  )}/users/${encodeURIComponent(userGuid)}/roles`;
  return request<UserRoleAssignment>('PUT', url, assignment);
};

export const deleteUserRoleAssignment = (
  tenantGuid: string,
  userGuid: string,
  assignmentGuid: string
) => {
  const url = `${getBaseUrl()}/v1.0/tenants/${encodeURIComponent(
    tenantGuid
  )}/users/${encodeURIComponent(userGuid)}/roles/${encodeURIComponent(assignmentGuid)}`;
  return request<void>('DELETE', url);
};

export const listCredentialScopeAssignments = (
  tenantGuid: string,
  credentialGuid: string,
  params: AuthorizationListParams = {}
) => {
  const url = `${getBaseUrl()}/v1.0/tenants/${encodeURIComponent(
    tenantGuid
  )}/credentials/${encodeURIComponent(credentialGuid)}/scopes${buildQuery({
    page: params.page ?? 0,
    pageSize: params.pageSize ?? 1000,
  })}`;
  return request<AuthorizationSearchResult<CredentialScopeAssignment>>('GET', url);
};

export const createCredentialScopeAssignment = (
  tenantGuid: string,
  credentialGuid: string,
  assignment: Partial<CredentialScopeAssignment>
) => {
  const url = `${getBaseUrl()}/v1.0/tenants/${encodeURIComponent(
    tenantGuid
  )}/credentials/${encodeURIComponent(credentialGuid)}/scopes`;
  return request<CredentialScopeAssignment>('PUT', url, assignment);
};

export const deleteCredentialScopeAssignment = (
  tenantGuid: string,
  credentialGuid: string,
  assignmentGuid: string
) => {
  const url = `${getBaseUrl()}/v1.0/tenants/${encodeURIComponent(
    tenantGuid
  )}/credentials/${encodeURIComponent(credentialGuid)}/scopes/${encodeURIComponent(
    assignmentGuid
  )}`;
  return request<void>('DELETE', url);
};

export const getUserEffectivePermissions = (
  tenantGuid: string,
  userGuid: string,
  graphGuid?: string
) => {
  const url = `${getBaseUrl()}/v1.0/tenants/${encodeURIComponent(
    tenantGuid
  )}/users/${encodeURIComponent(userGuid)}/permissions${buildQuery({ graphGuid })}`;
  return request<AuthorizationEffectivePermissionsResult>('GET', url);
};

export const getCredentialEffectivePermissions = (
  tenantGuid: string,
  credentialGuid: string,
  graphGuid?: string
) => {
  const url = `${getBaseUrl()}/v1.0/tenants/${encodeURIComponent(
    tenantGuid
  )}/credentials/${encodeURIComponent(credentialGuid)}/permissions${buildQuery({ graphGuid })}`;
  return request<AuthorizationEffectivePermissionsResult>('GET', url);
};
